### PR TITLE
Mailbox 273 multimailbox search support

### PR DESCRIPTION
**Pull request by rouazana**

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/LazyMessageSearchIndex.java [hunk 1]
```diff
@@ -20,16 +20,19 @@
 
 import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
```
Comment (aduprat) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/LazyMessageSearchIndex.java]: I think some imports don't belong to this commit.

File: mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java [hunk 2]
```diff
@@ -48,6 +44,8 @@
 import org.junit.Before;
 import org.junit.Test;
 
+import static org.assertj.core.api.Assertions.assertThat;
```
Comment (aduprat) [inline mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java]: static first

File: mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java [hunk 3]
```diff
@@ -126,580 +124,436 @@ public void setUp() throws Exception {
...
-        SearchQuery q2 = new SearchQuery();
-        q2.andCriteria(SearchQuery.bodyContains("body"));
-        Iterator<Long> it4 = index.search(null, mailbox, q2);
-        assertEquals(1L, it4.next().longValue());
-        assertEquals(2L, it4.next().longValue());
-        assertEquals(3L, it4.next().longValue());
-
-        assertFalse(it4.hasNext());
+    public void bodySearchShouldMatchSeveralEmails() throws Exception {    
+        SearchQuery q = new SearchQuery();
```
Comment (aduprat) [inline mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java]: Q? From James Bond?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            return this;
+        }
+        
+        public MultimailboxesSearchQuery build() {
+            return new MultimailboxesSearchQuery(searchQuery, mailboxIds.build());
+        }
+        
+    }
+
+    private SearchQuery searchQuery;
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: final

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    private ImmutableSet<MailboxId> mailboxIds;
+
+    public MultimailboxesSearchQuery(SearchQuery searchQuery, ImmutableSet<MailboxId> mailboxIds) {
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: private

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.model;
+
+import java.util.Arrays;
+
+import com.google.common.collect.ImmutableSet;
+
+public class MultimailboxesSearchQuery {
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: No tests?

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        private final SearchQuery searchQuery;
+        private ImmutableSet.Builder<MailboxId> mailboxIds;
+
+        private Builder(SearchQuery searchQuery) {
+            this.searchQuery = searchQuery;
+            this.mailboxIds = ImmutableSet.builder();
+        }
+        
+        public Builder inMailboxes(MailboxId... mailboxIds) {
+            this.mailboxIds.addAll(Arrays.asList(mailboxIds));
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: This may raise a NPE

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            return this;
+        }
+        
+        public MultimailboxesSearchQuery build() {
+            return new MultimailboxesSearchQuery(searchQuery, mailboxIds.build());
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: You may build a query without any mailbox, do we want to handle this choice?

File: server/container/guice/guice-common/src/main/java/org/apache/james/jmap/JMAPModule.java [hunk 9]
```diff
@@ -106,6 +106,8 @@ public void initModule() {
                     "MOVE support in MailboxManager is required by JMAP Module");
             Preconditions.checkArgument(mailboxManager.getSupportedMessageCapabilities().contains(MailboxManager.MessageCapabilities.Attachment),
                     "Attachment support in MailboxManager is required by JMAP Module");
+            Preconditions.checkArgument(mailboxManager.getSupportedSearchCapabilities().contains(MailboxManager.SearchCapabilities.MultimailboxSearch),
```
Comment (aduprat) [inline server/container/guice/guice-common/src/main/java/org/apache/james/jmap/JMAPModule.java]: This is a MessageSearchIndex capability.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/LazyMessageSearchIndex.java (hunk 1 above)
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/LazyMessageSearchIndex.java]: ok

File: mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java (hunk 2 above)
Comment (rouazana) [inline mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java]: ok

File: mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java (hunk 3 above)
Comment (rouazana) [inline mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java]: ok...

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java (hunk 6 above)
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: ok

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java (hunk 7 above)
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: nope, see tests :)

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java (hunk 8 above)
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: yes, that's all the purpose of this

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java (hunk 4 above)
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: ok

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java (hunk 5 above)
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: ok

File: server/container/guice/guice-common/src/main/java/org/apache/james/jmap/JMAPModule.java (hunk 9 above)
Comment (rouazana) [inline server/container/guice/guice-common/src/main/java/org/apache/james/jmap/JMAPModule.java]: We thought about that with matthieu, and decided to put it in mailboxManager for the same reason than message capabilities are in mailboxManager instead of messageManager: MessageSearchIndex like MessageManager need a MailboxManager.

Comment (aduprat) [inline server/container/guice/guice-common/src/main/java/org/apache/james/jmap/JMAPModule.java]: I don't think, as:
- MessageSearchIndex is not really linked to MailboxManager, not as MessageManager
- MessageSearchIndex as it's own configuration file `indexer.xml`
- MessageSearchIndex implementation is not coupled to MailboxManager implementation, so you can have James running on JPA with ES

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java [hunk 10]
```diff
@@ -45,15 +47,17 @@ public Builder inMailboxes(MailboxId... mailboxIds) {
         }
         
         public MultimailboxesSearchQuery build() {
+            Preconditions.checkNotNull(searchQuery);
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: checkState

Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MultimailboxesSearchQuery.java]: as it comes from the constructor, it's a checkNotNull but earlier.

**Comment by mbaechler** [comment]
up-to-date PR here : https://github.com/linagora/james-project/pull/347
